### PR TITLE
Issue #6877: Exclude files from code coverage

### PR DESCRIPTION
**Pull request by kailin0512**
## Proposed changes

Implement Exclude feature in [#6877](https://github.com/webdriverio/webdriverio/issues/6877)

## Types of changes

[//]: # (What types of changes does your code introduce to WebdriverIO?)
[//]: # (_Put an `x` in the boxes that apply_)

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update

## Checklist

[//]: # (_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._)

- [x] I have read the [CONTRIBUTING](https://github.com/webdriverio/webdriverio/blob/main/CONTRIBUTING.md) doc
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [x] I have added proper type definitions for new commands (if appropriate)

## Further comments

[//]: # (If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...)

### CODE REVIEW

**Comment by kailin0512** [comment]
Please review and let me know if some documents need to be added, thanks

File: packages/wdio-devtools-service/src/types.ts [hunk 1]
```diff
@@ -25,7 +25,11 @@ export interface CoverageReporterOptions {
     /**
      * Options for coverage report
      */
-    options?: any
+    options?: any,
+    /**
+     * Exclude code coverage files
+     */
+     exclude?: any[];
```
Comment (christian-bromann) [inline packages/wdio-devtools-service/src/types.ts:32]: Let's type it properly:
```suggestion
     * Exclude code coverage files
     * @default []
     */
    exclude?: (RegExp | string)[]
```

File: packages/wdio-devtools-service/src/types.ts [hunk 2]
```diff
@@ -25,7 +25,11 @@ export interface CoverageReporterOptions {
     /**
      * Options for coverage report
      */
-    options?: any
+    options?: any,
```
Comment (christian-bromann) [inline packages/wdio-devtools-service/src/types.ts:28]:
```suggestion
    options?: any
```

File: packages/wdio-devtools-service/src/gatherer/coverage.ts [hunk 3]
```diff
@@ -57,7 +57,24 @@ export default class CoverageGatherer extends EventEmitter {
         /**
          * continue with requests that aren't JS files
          */
+        var skipCoverageFlag = false;
```
Comment (christian-bromann) [inline packages/wdio-devtools-service/src/gatherer/coverage.ts:60]:
```suggestion
        var skipCoverageFlag = false
```

File: packages/wdio-devtools-service/src/gatherer/coverage.ts [hunk 4]
```diff
@@ -57,7 +57,24 @@ export default class CoverageGatherer extends EventEmitter {
...
         if (!request.url.endsWith('.js')) {
+            skipCoverageFlag = true;
```
Comment (christian-bromann) [inline packages/wdio-devtools-service/src/gatherer/coverage.ts:62]:
```suggestion
            skipCoverageFlag = true
```

File: packages/wdio-devtools-service/src/gatherer/coverage.ts [hunk 5]
```diff
@@ -57,7 +57,24 @@ export default class CoverageGatherer extends EventEmitter {
...
+        }
+
+        /**
+         * continue with requests that are part of exclude patterns
+         */
+        if(this._options.exclude){
+            for(const excludeFile of this._options.exclude){
+                if(request.url.match(excludeFile)){
+                    skipCoverageFlag = true;
```
Comment (christian-bromann) [inline packages/wdio-devtools-service/src/gatherer/coverage.ts:71]:
```suggestion
                    skipCoverageFlag = true
```

File: packages/wdio-devtools-service/src/gatherer/coverage.ts [hunk 6]
```diff
@@ -57,7 +57,24 @@ export default class CoverageGatherer extends EventEmitter {
...
+                    break;
```
Comment (christian-bromann) [inline packages/wdio-devtools-service/src/gatherer/coverage.ts:72]:
```suggestion
                    break
```

**Review by christian-bromann — COMMENTED** [review]
A couple of minor things (some of the code styling should be enfored)

**Comment by kailin0512** [comment]
Thanks, I will run lint in future contributions

**Comment by christian-bromann** [comment]
@kailin0512 thanks! Last thing: would you mind to sign the CLA?

**Comment by kailin0512** [comment]
> @kailin0512 thanks! Last thing: would you mind to sign the CLA?



> @kailin0512 thanks! Last thing: would you mind to sign the CLA?

Please let me check with my manager, thanks!

File: packages/wdio-devtools-service/src/gatherer/coverage.ts [hunk 7]
```diff
@@ -57,7 +57,24 @@ export default class CoverageGatherer extends EventEmitter {
         /**
          * continue with requests that aren't JS files
          */
+        var skipCoverageFlag = false
```
Comment (christian-bromann) [inline packages/wdio-devtools-service/src/gatherer/coverage.ts:60]:
```suggestion
        let skipCoverageFlag = false
```

**Comment by kailin0512** [comment]
@christian-bromann Do you have some thoughts why this unit test fails? I feel like it should be same with file ends with .html right?
https://github.com/webdriverio/webdriverio/pull/6920/checks?check_run_id=2679402411#step:7:1214

**Comment by christian-bromann** [comment]
@kailin0512 do these tests pass locally?

**Comment by kailin0512** [comment]
Not really, I run in local and it also seems to fail, I was wondering why babelTransform is called once? I feel like it is almost exact replication for file not ends with js (Like foo.html)

**Comment by kailin0512** [comment]
It also seems to me that these lines has not been covered, do you have some suggestions on why this happen or what I can do to add coverage? 
https://github.com/webdriverio/webdriverio/blob/main/packages/wdio-devtools-service/src/gatherer/coverage.ts#L73-L114

**Comment by kailin0512** [comment]
@christian-bromann It seems that e2e tests has broken
`expect(await browser.getElementProperty(elem[ELEMENT_KEY], 'value'))
            .toBe('Dropped!')`

**Comment by christian-bromann** [comment]
> It seems that e2e tests has broken

They were broken due to changes to the test pages. I made the test more stable and pushed a fix to `main`. Could you update your branch?

**Comment by christian-bromann** [comment]
Actually the `main` branch is still broken due to a different issue now. I am investigating.

**Comment by kailin0512** [comment]
Let me know when I can merge master branch and rerun the pipelines, thanks

**Comment by christian-bromann** [comment]
@kailin0512 it seems that the `main` branch is stable again (though e2e tests still seem to be very flaky due to unknown reasons). You can try to update your branch now.

**Review by christian-bromann — APPROVED** [review]
I will go ahead and merge this, as I want to make the next release. If any tests should fail I can fix them in `main`. 

Thank you for this very useful addition to the test coverage feature 👍

**Pull request merged**